internal/node/indexer/server.go

```diff
@@ -183,27 +216,74 @@ func (s *Server) handleTasks(ctx context.Context, tasks *engine.Tasks) error {
 	return nil
 }
 
-func (s *Server) initializeMeter() (err error) {
-	// init ethereum client
+// checkWorkerStatus checks if the worker is ready.
+func (s *Server) checkWorkerStatus(ctx context.Context, checkpoint engine.Checkpoint) (err error) {
+	// Check if the worker is ready
+	// Get the current block height/block number from the checkpoint state.
+	type CheckpointState struct {
+		BlockHeight uint64 `json:"block_height"`
+		BlockNumber uint64 `json:"block_number"`
+		EventID     uint64 `json:"event_id"`
+	}
+
+	var state CheckpointState
+	if err := json.Unmarshal(checkpoint.State, &state); err != nil {
+		zap.L().Error("unmarshal checkpoint state", zap.Error(err))
+		return err
+	}
+
 	switch s.config.Network {
 	case network.Arweave:
-		arweaveClient, err := arweave.NewClient()
+		latestHeight, err := s.arweaveClient.GetBlockHeight(ctx)
 		if err != nil {
-			return fmt.Errorf("new arweave client: %w", err)
+			zap.L().Error("get latest block height", zap.Error(err))
+			return err
 		}
 
-		s.arweaveClient = arweaveClient
+		currentHeight := state.BlockHeight
+
+		if uint64(latestHeight)-currentHeight < 100 {
+			// Cache worker status to Redis.
+			if err := s.updateWorkerStatus(ctx, s.config.Network.String(), s.config.Worker.String(), workerx.StatusReady.String()); err != nil {
+				return fmt.Errorf("cache token metadata: %w", err)
+			}
+		} else {
+			// Cache worker status to Redis.
+			if err := s.updateWorkerStatus(ctx, s.config.Network.String(), s.config.Worker.String(), workerx.StatusIndexing.String()); err != nil {
+				return fmt.Errorf("cache token metadata: %w", err)
+			}
+		}
 	case network.Farcaster:
+		// Check Timestamp
 		break
 	default:
-		ethereumClient, err := ethereum.Dial(context.Background(), s.config.Endpoint)
+		latestBlock, err := s.ethereumClient.BlockNumber(ctx)
 		if err != nil {
-			return fmt.Errorf("dial ethereum: %w", err)
+			zap.L().Error("get latest block number", zap.Error(err))
+			return err
 		}
 
-		s.ethereumClient = ethereumClient
+		latestBlockNumber := latestBlock.Int64()
+
+		currentBlockNumber := state.BlockNumber
+
+		if uint64(latestBlockNumber)-currentBlockNumber < 100 {
+			// Cache worker status to Redis.
+			if err := s.updateWorkerStatus(ctx, s.config.Network.String(), s.config.Worker.String(), workerx.StatusReady.String()); err != nil {
+				return fmt.Errorf("cache token metadata: %w", err)
+			}
+		} else {
+			// Cache worker status to Redis.
+			if err := s.updateWorkerStatus(ctx, s.config.Network.String(), s.config.Worker.String(), workerx.StatusIndexing.String()); err != nil {
+				return fmt.Errorf("cache token metadata: %w", err)
+			}
+		}
```

Comment:
once again, the switch cases have so many repeated logics that can be abstracted.